File: core/src/commonMain/kotlin/com/bluzi/babymonitor/xiaomi/Crypto.kt
package com.bluzi.babymonitor.xiaomi

import com.bluzi.babymonitor.platform.secureRandomBytes
import com.bluzi.babymonitor.platform.wallClockMs

// Kotlin port of c100/src/xiaomi/crypto.ts (itself a port of go2rtc's cloud.go / miss crypto).
// Every function here is pure and covered by interop vectors in protocol-vectors.json.

object Crypto {
    fun md5(data: ByteArray): ByteArray = md5Digest(data)
    fun sha1(data: ByteArray): ByteArray = sha1Digest(data)
    fun sha256(data: ByteArray): ByteArray = sha256Digest(data)

    fun randomBytes(n: Int): ByteArray = secureRandomBytes(n)

    fun randString(n: Int): String {
        val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
        return buildString(n) {
            for (b in randomBytes(n)) append(alphabet[(b.toInt() and 0xff) % alphabet.length])
        }
    }

    /** PROTO-1: passwords travel as uppercase-hex MD5. */
    fun passwordHash(password: String): String =
        md5(password.encodeToByteArray()).toHex().uppercase()

    // PROTO-8: RC4 with the first 1024 keystream bytes discarded.
    fun rc4(key: ByteArray, data: ByteArray, discard: Int = 1024): ByteArray {
        val s = IntArray(256) { it }
        var j = 0
        for (i in 0 until 256) {
            j = (j + s[i] + (key[i % key.size].toInt() and 0xff)) and 0xff
            val t = s[i]; s[i] = s[j]; s[j] = t
        }
        var i = 0
        j = 0
        repeat(discard) {
            i = (i + 1) and 0xff
            j = (j + s[i]) and 0xff
            val t = s[i]; s[i] = s[j]; s[j] = t
        }
        val out = ByteArray(data.size)
        for (n in data.indices) {
            i = (i + 1) and 0xff
            j = (j + s[i]) and 0xff
            val t = s[i]; s[i] = s[j]; s[j] = t
            out[n] = (data[n].toInt() xor s[(s[i] + s[j]) and 0xff]).toByte()
        }
        return out
    }

    // --- ChaCha20 (RFC 8439 core, counter starts at 0) -------------------------

    private fun chachaBlock(key: IntArray, counter: Int, nonce: IntArray, out: ByteArray, outOff: Int) {
        val st = IntArray(16)
        st[0] = 0x61707865; st[1] = 0x3320646e; st[2] = 0x79622d32; st[3] = 0x6b206574
        key.copyInto(st, 4)
        st[12] = counter
        nonce.copyInto(st, 13)
        val x = st.copyOf()
        repeat(10) {
            qr(x, 0, 4, 8, 12); qr(x, 1, 5, 9, 13); qr(x, 2, 6, 10, 14); qr(x, 3, 7, 11, 15)
            qr(x, 0, 5, 10, 15); qr(x, 1, 6, 11, 12); qr(x, 2, 7, 8, 13); qr(x, 3, 4, 9, 14)
        }
        for (k in 0 until 16) {
            val w = x[k] + st[k]
            val base = outOff + k * 4
            out[base] = w.toByte()
            out[base + 1] = (w ushr 8).toByte()
            out[base + 2] = (w ushr 16).toByte()
            out[base + 3] = (w ushr 24).toByte()
        }
    }

    private fun qr(x: IntArray, a: Int, b: Int, c: Int, d: Int) {
        x[a] += x[b]; x[d] = (x[d] xor x[a]).rotateLeft(16)
        x[c] += x[d]; x[b] = (x[b] xor x[c]).rotateLeft(12)
        x[a] += x[b]; x[d] = (x[d] xor x[a]).rotateLeft(8)
        x[c] += x[d]; x[b] = (x[b] xor x[c]).rotateLeft(7)
    }

    private fun leWords(bytes: ByteArray, off: Int, count: Int): IntArray =
        IntArray(count) { i ->
            val b = off + i * 4
            (bytes[b].toInt() and 0xff) or
                ((bytes[b + 1].toInt() and 0xff) shl 8) or
                ((bytes[b + 2].toInt() and 0xff) shl 16) or
                ((bytes[b + 3].toInt() and 0xff) shl 24)
        }

    fun chacha20Xor(key32: ByteArray, nonce12: ByteArray, data: ByteArray): ByteArray {
        require(key32.size == 32) { "chacha: key must be 32 bytes" }
        require(nonce12.size == 12) { "chacha: nonce must be 12 bytes" }
        val keyWords = leWords(key32, 0, 8)
        val nonceWords = leWords(nonce12, 0, 3)
        val out = ByteArray(data.size)
        val block = ByteArray(64)
        var counter = 0
        var off = 0
        while (off < data.size) {
            chachaBlock(keyWords, counter, nonceWords, block, 0)
            val n = minOf(64, data.size - off)
            for (k in 0 until n) out[off + k] = (data[off + k].toInt() xor block[k].toInt()).toByte()
            counter++
            off += n
        }
        return out
    }

    /** PROTO-14: wire format is [8 nonce bytes][chacha20(key, 0x00000000‖nonce8)(plaintext)]. */
    fun chachaEncode(plaintext: ByteArray, key32: ByteArray): ByteArray {
        val nonce8 = randomBytes(8)
        return concatBytes(nonce8, chachaDecodeNonce(plaintext, nonce8, key32))
    }

    fun chachaDecode(src: ByteArray, key32: ByteArray): ByteArray {
        require(src.size >= 8) { "miss: payload too short" }
        return chachaDecodeNonce(src.copyOfRange(8, src.size), src.copyOfRange(0, 8), key32)
    }

    fun chachaDecodeNonce(src: ByteArray, nonce8: ByteArray, key32: ByteArray): ByteArray {
        val nonce12 = ByteArray(12)
        nonce8.copyInto(nonce12, 4)
        return chacha20Xor(key32, nonce12, src)
    }

    // --- NaCl box precomputation (PROTO-13) -------------------------------------
    // box.before(pub, priv) = HSalsa20(key = X25519(priv, pub), input = 0^16)

    fun generateBoxKeyPair(): Pair<ByteArray, ByteArray> {
        val priv = randomBytes(32)
        return x25519ScalarMultBase(priv) to priv
    }

    fun calcSharedKey(devicePublicHex: String, clientPrivateHex: String): ByteArray {
        val pub = devicePublicHex.hexToBytes()
        val priv = clientPrivateHex.hexToBytes()
        require(pub.size == 32 && priv.size == 32) { "box key must be 32 bytes" }
        return hsalsa20(x25519ScalarMult(priv, pub))
    }

    /** HSalsa20 with the "expand 32-byte k" constants and a zero input block. */
    internal fun hsalsa20(key32: ByteArray): ByteArray {
        val k = leWords(key32, 0, 8)
        val x = IntArray(16)
        x[0] = 0x61707865; x[5] = 0x3320646e; x[10] = 0x79622d32; x[15] = 0x6b206574
        x[1] = k[0]; x[2] = k[1]; x[3] = k[2]; x[4] = k[3]
        x[11] = k[4]; x[12] = k[5]; x[13] = k[6]; x[14] = k[7]
        // x[6..9] = input = zeros
        repeat(10) {
            // column round
            x[4] = x[4] xor (x[0] + x[12]).rotateLeft(7)
            x[8] = x[8] xor (x[4] + x[0]).rotateLeft(9)
            x[12] = x[12] xor (x[8] + x[4]).rotateLeft(13)
            x[0] = x[0] xor (x[12] + x[8]).rotateLeft(18)
            x[9] = x[9] xor (x[5] + x[1]).rotateLeft(7)
            x[13] = x[13] xor (x[9] + x[5]).rotateLeft(9)
            x[1] = x[1] xor (x[13] + x[9]).rotateLeft(13)
            x[5] = x[5] xor (x[1] + x[13]).rotateLeft(18)
            x[14] = x[14] xor (x[10] + x[6]).rotateLeft(7)
            x[2] = x[2] xor (x[14] + x[10]).rotateLeft(9)
            x[6] = x[6] xor (x[2] + x[14]).rotateLeft(13)
            x[10] = x[10] xor (x[6] + x[2]).rotateLeft(18)
            x[3] = x[3] xor (x[15] + x[11]).rotateLeft(7)
            x[7] = x[7] xor (x[3] + x[15]).rotateLeft(9)
            x[11] = x[11] xor (x[7] + x[3]).rotateLeft(13)
            x[15] = x[15] xor (x[11] + x[7]).rotateLeft(18)
            // row round
            x[1] = x[1] xor (x[0] + x[3]).rotateLeft(7)
            x[2] = x[2] xor (x[1] + x[0]).rotateLeft(9)
            x[3] = x[3] xor (x[2] + x[1]).rotateLeft(13)
            x[0] = x[0] xor (x[3] + x[2]).rotateLeft(18)
            x[6] = x[6] xor (x[5] + x[4]).rotateLeft(7)
            x[7] = x[7] xor (x[6] + x[5]).rotateLeft(9)
            x[4] = x[4] xor (x[7] + x[6]).rotateLeft(13)
            x[5] = x[5] xor (x[4] + x[7]).rotateLeft(18)
            x[11] = x[11] xor (x[10] + x[9]).rotateLeft(7)
            x[8] = x[8] xor (x[11] + x[10]).rotateLeft(9)
            x[9] = x[9] xor (x[8] + x[11]).rotateLeft(13)
            x[10] = x[10] xor (x[9] + x[8]).rotateLeft(18)
            x[12] = x[12] xor (x[15] + x[14]).rotateLeft(7)
            x[13] = x[13] xor (x[12] + x[15]).rotateLeft(9)
            x[14] = x[14] xor (x[13] + x[12]).rotateLeft(13)
            x[15] = x[15] xor (x[14] + x[13]).rotateLeft(18)
        }
        val outWords = intArrayOf(x[0], x[5], x[10], x[15], x[6], x[7], x[8], x[9])
        val out = ByteArray(32)
        for (i in 0 until 8) {
            out[i * 4] = outWords[i].toByte()
            out[i * 4 + 1] = (outWords[i] ushr 8).toByte()
            out[i * 4 + 2] = (outWords[i] ushr 16).toByte()
            out[i * 4 + 3] = (outWords[i] ushr 24).toByte()
        }
        return out
    }

    // --- Mi cloud signed-request bits (PROTO-7) ---------------------------------

    /** 8 random bytes + big-endian u32 minutes-since-epoch. */
    fun genNonce(nowMs: Long = wallClockMs(), random8: ByteArray = randomBytes(8)): ByteArray {
        require(random8.size == 8)
        val nonce = ByteArray(12)
        random8.copyInto(nonce, 0)
        nonce.putBeU32(8, nowMs / 1000 / 60)
        return nonce
    }

    fun genSignedNonce(ssecurity: ByteArray, nonce: ByteArray): ByteArray =
        sha256(concatBytes(ssecurity, nonce))

    fun genSignatureB64(method: String, path: String, data: String, rc4Hash: String?, signedNonce: ByteArray): String {
        var s = "$method&$path&data=$data"
        if (rc4Hash != null) s += "&rc4_hash__=$rc4Hash"
        s += "&" + signedNonce.toBase64()
        return sha1(s.encodeToByteArray()).toBase64()
    }
}

File: windows/src/BabyMonitor.Core/Xiaomi/Crypto.cs
using System.Security.Cryptography;
using System.Text;
using BabyMonitor.Core.Platform;

namespace BabyMonitor.Core.Xiaomi;

/// <summary>
/// The wire's cryptography — port of the Kotlin core's <c>xiaomi/Crypto.kt</c> (itself a port of
/// c100's crypto.ts, itself go2rtc's). Every function here is pure and covered by the interop
/// vectors in protocol-vectors.json.
///
/// MD5/SHA-1/SHA-256 come from the framework rather than being ported by hand. The Kotlin core wrote
/// its own because Kotlin/Native has no MessageDigest; .NET has had these for twenty years, and a
/// hand-rolled hash next to a correct one in the box is a liability, not an asset. The signing scheme
/// is still proven end to end by the vectors — which is what actually guards it.
/// </summary>
public static class Crypto
{
    private const string RandomAlphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

    public static byte[] Md5(byte[] data) => MD5.HashData(data);

    public static byte[] Sha1(byte[] data) => SHA1.HashData(data);

    public static byte[] Sha256(byte[] data) => SHA256.HashData(data);

    public static byte[] RandomBytes(int n) => Clock.SecureRandomBytes(n);

    public static string RandString(int n)
    {
        var sb = new StringBuilder(n);
        foreach (var b in RandomBytes(n))
        {
            sb.Append(RandomAlphabet[b % RandomAlphabet.Length]);
        }

        return sb.ToString();
    }

    /// <summary>PROTO-1: passwords travel as uppercase-hex MD5.</summary>
    public static string PasswordHash(string password) =>
        Md5(Encoding.UTF8.GetBytes(password)).ToHex().ToUpperInvariant();

    /// <summary>PROTO-8: RC4 with the first 1024 keystream bytes discarded.</summary>
    public static byte[] Rc4(byte[] key, byte[] data, int discard = 1024)
    {
        var s = new int[256];
        for (var i = 0; i < 256; i++)
        {
            s[i] = i;
        }

        var j = 0;
        for (var i = 0; i < 256; i++)
        {
            j = (j + s[i] + key[i % key.Length]) & 0xff;
            (s[i], s[j]) = (s[j], s[i]);
        }

        var a = 0;
        j = 0;
        for (var n = 0; n < discard; n++)
        {
            a = (a + 1) & 0xff;
            j = (j + s[a]) & 0xff;
            (s[a], s[j]) = (s[j], s[a]);
        }

        var out_ = new byte[data.Length];
        for (var n = 0; n < data.Length; n++)
        {
            a = (a + 1) & 0xff;
            j = (j + s[a]) & 0xff;
            (s[a], s[j]) = (s[j], s[a]);
            out_[n] = (byte)(data[n] ^ s[(s[a] + s[j]) & 0xff]);
        }

        return out_;
    }

    // --- ChaCha20 (RFC 8439 core, counter starts at 0) --------------------------

    public static byte[] Chacha20Xor(byte[] key32, byte[] nonce12, byte[] data)
    {
        if (key32.Length != 32)
        {
            throw new ArgumentException("chacha: key must be 32 bytes", nameof(key32));
        }

        if (nonce12.Length != 12)
        {
            throw new ArgumentException("chacha: nonce must be 12 bytes", nameof(nonce12));
        }

        var keyWords = LeWords(key32, 0, 8);
        var nonceWords = LeWords(nonce12, 0, 3);
        var out_ = new byte[data.Length];
        var block = new byte[64];
        var counter = 0;
        var off = 0;
        while (off < data.Length)
        {
            ChachaBlock(keyWords, counter, nonceWords, block);
            var n = Math.Min(64, data.Length - off);
            for (var k = 0; k < n; k++)
            {
                out_[off + k] = (byte)(data[off + k] ^ block[k]);
            }

            counter++;
            off += n;
        }

        return out_;
    }

    /// <summary>PROTO-14: wire format is [8 nonce bytes][chacha20(key, 0x00000000‖nonce8)(plaintext)].</summary>
    public static byte[] ChachaEncode(byte[] plaintext, byte[] key32)
    {
        var nonce8 = RandomBytes(8);
        return Bytes.Concat(nonce8, ChachaDecodeNonce(plaintext, nonce8, key32));
    }

    public static byte[] ChachaDecode(byte[] src, byte[] key32)
    {
        if (src.Length < 8)
        {
            throw new ArgumentException("miss: payload too short", nameof(src));
        }

        return ChachaDecodeNonce(src.Slice(8, src.Length), src.Slice(0, 8), key32);
    }

    public static byte[] ChachaDecodeNonce(byte[] src, byte[] nonce8, byte[] key32)
    {
        var nonce12 = new byte[12];
        Buffer.BlockCopy(nonce8, 0, nonce12, 4, 8);
        return Chacha20Xor(key32, nonce12, src);
    }

    // --- NaCl box precomputation (PROTO-13) -------------------------------------
    // box.before(pub, priv) = HSalsa20(key = X25519(priv, pub), input = 0^16)

    public static (byte[] Public, byte[] Private) GenerateBoxKeyPair()
    {
        var priv = RandomBytes(32);
        return (Curve25519.ScalarMultBase(priv), priv);
    }

    public static byte[] CalcSharedKey(string devicePublicHex, string clientPrivateHex)
    {
        var pub = devicePublicHex.HexToBytes();
        var priv = clientPrivateHex.HexToBytes();
        if (pub.Length != 32 || priv.Length != 32)
        {
            throw new ArgumentException("box key must be 32 bytes");
        }

        return Hsalsa20(Curve25519.ScalarMult(priv, pub));
    }

    /// <summary>HSalsa20 with the "expand 32-byte k" constants and a zero input block.</summary>
    public static byte[] Hsalsa20(byte[] key32)
    {
        var k = LeWords(key32, 0, 8);
        var x = new uint[16];
        x[0] = 0x61707865;
        x[5] = 0x3320646e;
        x[10] = 0x79622d32;
        x[15] = 0x6b206574;
        x[1] = k[0];
        x[2] = k[1];
        x[3] = k[2];
        x[4] = k[3];
        x[11] = k[4];
        x[12] = k[5];
        x[13] = k[6];
        x[14] = k[7];
        // x[6..9] = input = zeros

        for (var round = 0; round < 10; round++)
        {
            // column round
            x[4] ^= Rotl(x[0] + x[12], 7);
            x[8] ^= Rotl(x[4] + x[0], 9);
            x[12] ^= Rotl(x[8] + x[4], 13);
            x[0] ^= Rotl(x[12] + x[8], 18);
            x[9] ^= Rotl(x[5] + x[1], 7);
            x[13] ^= Rotl(x[9] + x[5], 9);
            x[1] ^= Rotl(x[13] + x[9], 13);
            x[5] ^= Rotl(x[1] + x[13], 18);
            x[14] ^= Rotl(x[10] + x[6], 7);
            x[2] ^= Rotl(x[14] + x[10], 9);
            x[6] ^= Rotl(x[2] + x[14], 13);
            x[10] ^= Rotl(x[6] + x[2], 18);
            x[3] ^= Rotl(x[15] + x[11], 7);
            x[7] ^= Rotl(x[3] + x[15], 9);
            x[11] ^= Rotl(x[7] + x[3], 13);
            x[15] ^= Rotl(x[11] + x[7], 18);

            // row round
            x[1] ^= Rotl(x[0] + x[3], 7);
            x[2] ^= Rotl(x[1] + x[0], 9);
            x[3] ^= Rotl(x[2] + x[1], 13);
            x[0] ^= Rotl(x[3] + x[2], 18);
            x[6] ^= Rotl(x[5] + x[4], 7);
            x[7] ^= Rotl(x[6] + x[5], 9);
            x[4] ^= Rotl(x[7] + x[6], 13);
            x[5] ^= Rotl(x[4] + x[7], 18);
            x[11] ^= Rotl(x[10] + x[9], 7);
            x[8] ^= Rotl(x[11] + x[10], 9);
            x[9] ^= Rotl(x[8] + x[11], 13);
            x[10] ^= Rotl(x[9] + x[8], 18);
            x[12] ^= Rotl(x[15] + x[14], 7);
            x[13] ^= Rotl(x[12] + x[15], 9);
            x[14] ^= Rotl(x[13] + x[12], 13);
            x[15] ^= Rotl(x[14] + x[13], 18);
        }

        var outWords = new[] { x[0], x[5], x[10], x[15], x[6], x[7], x[8], x[9] };
        var out_ = new byte[32];
        for (var i = 0; i < 8; i++)
        {
            out_[i * 4] = (byte)outWords[i];
            out_[(i * 4) + 1] = (byte)(outWords[i] >> 8);
            out_[(i * 4) + 2] = (byte)(outWords[i] >> 16);
            out_[(i * 4) + 3] = (byte)(outWords[i] >> 24);
        }

        return out_;
    }

    // --- Mi cloud signed-request bits (PROTO-7) ---------------------------------

    /// <summary>8 random bytes + big-endian u32 minutes-since-epoch.</summary>
    public static byte[] GenNonce(long? nowMs = null, byte[]? random8 = null)
    {
        var stamp = nowMs ?? Clock.WallClockMs();
        var random = random8 ?? RandomBytes(8);
        if (random.Length != 8)
        {
            throw new ArgumentException("nonce: random part must be 8 bytes", nameof(random8));
        }

        var nonce = new byte[12];
        Buffer.BlockCopy(random, 0, nonce, 0, 8);
        nonce.PutBeU32(8, stamp / 1000 / 60);
        return nonce;
    }

    public static byte[] GenSignedNonce(byte[] ssecurity, byte[] nonce) =>
        Sha256(Bytes.Concat(ssecurity, nonce));

    public static string GenSignatureB64(string method, string path, string data, string? rc4Hash, byte[] signedNonce)
    {
        var s = $"{method}&{path}&data={data}";
        if (rc4Hash != null)
        {
            s += $"&rc4_hash__={rc4Hash}";
        }

        s += "&" + signedNonce.ToBase64();
        return Sha1(Encoding.UTF8.GetBytes(s)).ToBase64();
    }

    private static void ChachaBlock(uint[] key, int counter, uint[] nonce, byte[] out_)
    {
        var st = new uint[16];
        st[0] = 0x61707865;
        st[1] = 0x3320646e;
        st[2] = 0x79622d32;
        st[3] = 0x6b206574;
        Array.Copy(key, 0, st, 4, 8);
        st[12] = (uint)counter;
        Array.Copy(nonce, 0, st, 13, 3);

        var x = (uint[])st.Clone();
        for (var round = 0; round < 10; round++)
        {
            Qr(x, 0, 4, 8, 12);
            Qr(x, 1, 5, 9, 13);
            Qr(x, 2, 6, 10, 14);
            Qr(x, 3, 7, 11, 15);
            Qr(x, 0, 5, 10, 15);
            Qr(x, 1, 6, 11, 12);
            Qr(x, 2, 7, 8, 13);
            Qr(x, 3, 4, 9, 14);
        }

        for (var k = 0; k < 16; k++)
        {
            var w = x[k] + st[k];
            var b = k * 4;
            out_[b] = (byte)w;
            out_[b + 1] = (byte)(w >> 8);
            out_[b + 2] = (byte)(w >> 16);
            out_[b + 3] = (byte)(w >> 24);
        }
    }

    private static void Qr(uint[] x, int a, int b, int c, int d)
    {
        x[a] += x[b];
        x[d] = Rotl(x[d] ^ x[a], 16);
        x[c] += x[d];
        x[b] = Rotl(x[b] ^ x[c], 12);
        x[a] += x[b];
        x[d] = Rotl(x[d] ^ x[a], 8);
        x[c] += x[d];
        x[b] = Rotl(x[b] ^ x[c], 7);
    }

    private static uint Rotl(uint value, int bits) => (value << bits) | (value >> (32 - bits));

    private static uint[] LeWords(byte[] bytes, int off, int count)
    {
        var words = new uint[count];
        for (var i = 0; i < count; i++)
        {
            var b = off + (i * 4);
            words[i] = bytes[b] |
                ((uint)bytes[b + 1] << 8) |
                ((uint)bytes[b + 2] << 16) |
                ((uint)bytes[b + 3] << 24);
        }

        return words;
    }
}
